client/fs_endpoint.go

```diff
@@ -0,0 +1,908 @@
+	if limit <= 0 {
+		fileReader = file
+	} else {
+		fileReader = io.LimitReader(file, limit)
+	}
+
+	// Create a tomb to cancel watch events
+	waitCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Create a variable to allow setting the last event
+	var lastEvent string
+
+	// Only create the file change watcher once. But we need to do it after we
+	// read and reach EOF.
+	var changes *watch.FileChanges
+
+	// Start streaming the data
+	data := make([]byte, streamFrameSize)
+OUTER:
+	for {
+		// Read up to the max frame size
+		n, readErr := fileReader.Read(data)
+
+		// Update the offset
+		offset += int64(n)
+
+		// Return non-EOF errors
+		if readErr != nil && readErr != io.EOF {
+			return readErr
+		}
+
+		// Send the frame
+		if n != 0 || lastEvent != "" {
+			if err := framer.Send(path, lastEvent, data[:n], offset); err != nil {
+				return parseFramerErr(err)
+			}
+		}
+
+		// Clear the last event
+		if lastEvent != "" {
+			lastEvent = ""
+		}
+
+		// Just keep reading
+		if readErr == nil {
+			continue
+		}
+
+		// If EOF is hit, wait for a change to the file
+		if changes == nil {
+			changes, err = fs.ChangeEvents(waitCtx, path, offset)
+			if err != nil {
+				return err
+			}
+		}
+
+		for {
+			select {
+			case <-changes.Modified:
+				continue OUTER
+			case <-changes.Deleted:
+				return parseFramerErr(framer.Send(path, deleteEvent, nil, offset))
+			case <-changes.Truncated:
+				// Close the current reader
+				if err := file.Close(); err != nil {
+					return err
+				}
+
+				// Get a new reader at offset zero
+				offset = 0
+				var err error
+				file, err = fs.ReadAt(path, offset)
+				if err != nil {
+					return err
+				}
+				defer file.Close()
+
+				if limit <= 0 {
+					fileReader = file
+				} else {
+					// Get the current limit
+					lr, ok := fileReader.(*io.LimitedReader)
+					if !ok {
+						return fmt.Errorf("unable to determine remaining read limit")
+					}
+
+					fileReader = io.LimitReader(file, lr.N)
+				}
+
+				// Store the last event
+				lastEvent = truncateEvent
+				continue OUTER
+			case <-framer.ExitCh():
+				return parseFramerErr(framer.Err())
+			case <-ctx.Done():
+				return nil
+			case err, ok := <-eofCancelCh:
+				if !ok {
+					return nil
+				}
+
+				return err
+			}
+		}
+	}
+}
+
+// blockUntilNextLog returns a channel that will have data sent when the next
+// log index or anything greater is created.
+func blockUntilNextLog(ctx context.Context, fs allocdir.AllocDirFS, logPath, task, logType string, nextIndex int64) chan error {
+	nextPath := filepath.Join(logPath, fmt.Sprintf("%s.%s.%d", task, logType, nextIndex))
+	next := make(chan error, 1)
+
+	go func() {
+		eofCancelCh, err := fs.BlockUntilExists(ctx, nextPath)
+		if err != nil {
+			next <- err
+			close(next)
+			return
+		}
+
+		ticker := time.NewTicker(nextLogCheckRate)
+		defer ticker.Stop()
+		scanCh := ticker.C
+		for {
+			select {
+			case <-ctx.Done():
+				next <- nil
```

Comment:
You might use `fallthrough` to avoid duplication in each case statement.

Reply:
select statements don't allow fallthrough